agent/immortalstreams/backedpipe/backed_pipe.go

```diff
@@ -0,0 +1,346 @@
+	ErrInvalidSequenceNumber  = xerrors.New("remote sequence number exceeds local sequence")
+	ErrReconnectWriterFailed  = xerrors.New("reconnect writer failed")
+)
+
+// connectionState represents the current state of the BackedPipe connection.
+type connectionState int
+
+const (
+	// connected indicates the pipe is connected and operational.
+	connected connectionState = iota
+	// disconnected indicates the pipe is not connected but not closed.
+	disconnected
+	// reconnecting indicates a reconnection attempt is in progress.
+	reconnecting
+	// closed indicates the pipe is permanently closed.
+	closed
+)
+
+// ErrorEvent represents an error from a reader or writer with connection generation info.
+type ErrorEvent struct {
+	Err        error
+	Component  string // "reader" or "writer"
+	Generation uint64 // connection generation when error occurred
+}
+
+const (
+	// Default buffer capacity used by the writer - 64MB
+	DefaultBufferSize = 64 * 1024 * 1024
+)
+
+// Reconnector is an interface for establishing connections when the BackedPipe needs to reconnect.
+// Implementations should:
+// 1. Establish a new connection to the remote side
+// 2. Exchange sequence numbers with the remote side
+// 3. Return the new connection and the remote's current sequence number
+//
+// The writerSeqNum parameter is the local writer's current sequence number,
+// which should be sent to the remote side so it knows where to resume reading from.
+//
+// The returned readerSeqNum should be the remote side's current sequence number,
+// which indicates where the local reader should resume from.
+type Reconnector interface {
+	Reconnect(ctx context.Context, writerSeqNum uint64) (conn io.ReadWriteCloser, readerSeqNum uint64, err error)
+}
+
+// BackedPipe provides a reliable bidirectional byte stream over unreliable network connections.
+// It orchestrates a BackedReader and BackedWriter to provide transparent reconnection
+// and data replay capabilities.
+type BackedPipe struct {
+	ctx         context.Context
+	cancel      context.CancelFunc
+	mu          sync.RWMutex
+	reader      *BackedReader
+	writer      *BackedWriter
+	reconnector Reconnector
+	conn        io.ReadWriteCloser
+
+	// State machine
+	state   connectionState
+	connGen uint64 // Increments on each successful reconnection
+
+	// Unified error handling with generation filtering
+	errorChan chan ErrorEvent
+
+	// singleflight group to dedupe concurrent ForceReconnect calls
+	sf singleflight.Group
+
+	// Track first error per generation to avoid duplicate reconnections
+	lastErrorGen uint64
+}
+
+// NewBackedPipe creates a new BackedPipe with default options and the specified reconnector.
+// The pipe starts disconnected and must be connected using Connect().
+func NewBackedPipe(ctx context.Context, reconnector Reconnector) *BackedPipe {
+	pipeCtx, cancel := context.WithCancel(ctx)
+
+	errorChan := make(chan ErrorEvent, 10) // Buffered for async error reporting
```

Comment:
What's the deal with 10? Like, why is that the right size for this buffer?

In general, I'd encourage you to consider anything other than 0 or 1 to be suspicious. There are legit uses for other numbers, but generally only if we are "buffering" to smooth out spiky usage.